keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,575 @@
+previous example:
+```golang
+type Root struct {
+  // This is now a pointer, we can provide a default:
+  // +default={name: pointer-name}
+  Entry *SubLevel
+}
+
+type SubLevel struct {
+  // +defaut=default-name
+  Name string
+  // +default=0
+  Number int
+}
+```
+
+We would generate the following OpenAPI definition, both for CRD types (by kubebuilder)
+and built-in types:
+```yaml
+Root:
+  type: object
+  default: {}
+  properties:
+    entry:
+      $ref: SubLevel
+      default: {name: pointer-name}
+SubLevel:
+  type: object
+  properties:
+    name:
+      type: string
+      default: default-name
+    number:
+      type: integer
+      default: 0
+```
+
+which would have the following behavior, both for CRD types and built-in types:
+```python
+>>> default('null')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{}')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{entry: null}')
+{"entry": {"name": "pointer-name", "number": 0}}
+>>> default('{entry: {}}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+#### Non-pointer fields
+
+Non-pointer fields are also important, since they have a specific unmarshalling mechanism in Go.
+
+For CRDs, we can assume that invalid zero-values are omitted, otherwise
+the objects would technically be invalid.
+
+For built-ins, we can differentiate two cases:
+1. The zero-value for the field is invalid (e.g. if empty string is not a valid value)
+2. The zero-value is valid, but is also the default (e.g. enabled false by default).
+
+Both these cases will be handled properly, here's an example:
+```golang
+type Object struct {
+     // This field has an invalid zero-value, empty string is not a
+     // valid string.
+     // +default=default-name
+     Name string
+
+     // This field is defaulted to 0 with no defaulter.
+     // +default=0
+     Defaulted int
+}
+```
+
+This would generate the following OpenAPI:
+```yaml
+Object:
+  type: object
+  properties:
+    invalid:
+      type: string
+      default: default-name
+    defaulted:
+      type: boolean
+      default: false
+```
+
+Which, for both CRD types and built-in types, would generate the following behavior:
+```python
+>>> default('{}')
+{"entry": {"name": "default-name", "number": 0}}
+>>> default('{entry: {name: other-name}}')
+{"entry": {"name": "other-name", "number": 0}}
+```
+
+Note that the generated code for built-in will look like this:
+```golang
+func GenerateDefault(obj *Object) {
+     if obj.Name == "" {
+          obj.Name = "default-name"
+     }
+     if obj.Number == 0 {
+          obj.Number = 0
+     }
+}
+```
+
+### Risks and Mitigations
+
+The main risk is that we could end-up with different defaults for
+existing types which would be extremely confusing for our users. This
+can be mitigated by double-checking the existing tests and adding new
+tests for gap in coverage.
```

Comment:
That seems orthogonal. We haven't ever generated defaulting tests before. Ideally we test the `+default` mechanism extensively rather than testing how resources actually get defaulted.